### PR TITLE
Upgrade to sdk 47

### PR DESCRIPTION
**Pull request by faddat**
This is a new version of the upgrade to sdk 47.

I've squashed work from these paths:

go.mod
go.sum
third_party
proto
Makefile

and the rest should begin no different from main.  You'll note that in go.mod there are forked versions of cosmos-sdk and ibc-go.  Please let us know if it is best to:

* make a pr to ibc-go OR move files into this repo if possible

AND

* weather or not if it is possible to upgrade interchain-security to sdk-47 without the changes to the cosmos-sdk that you see in the fork in go.mod

This branch contains the work of five people over nearly a month.  Instead of closing the PR when asked for review, as @mpoke did yesterday, please provide guidance if anything isn't to your liking. 

We feel that it is undesirable to upgrade the repository in the manner described in #852 -- specifically there's little sense in upgrading just one piece of the repository.  Let's instead make this module path

```go
module github.com/cosmos/interchain-security/v2
```

And then test v2 against v1 because that is the standard manner of doing these kinds of things in Cosmos.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] Added `!` to the type prefix if API or client breaking change
* [ ] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] Provided a link to the relevant issue or specification
* [ ] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
* [ ] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry to `CHANGELOG.md`
* [ ] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [ ] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed all author checklist items have been addressed 
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

File: x/ccv/provider/client/proposal_handler.go [hunk 1]
```diff
@@ -403,13 +331,13 @@ func postConsumerRemovalProposalHandlerFn(clientCtx client.Context) http.Handler
...
+		//		}
 
-		if rest.CheckBadRequestError(w, msg.ValidateBasic()) {
-			return
-		}
+		//		if rest.CheckBadRequestError(w, msg.ValidateBasic()) {
+		//			return
+		//		}
 
 		tx.WriteGeneratedTxResponse(clientCtx, w, req.BaseReq, msg)
```
Comment (faddat) [inline x/ccv/provider/client/proposal_handler.go:342]: @sontrinh16 @pysel

hey guys could you please fix up this file?

Comment (sontrinh16) [reply x/ccv/provider/client/proposal_handler.go]: yes i have remove all the rest usage let me clean up the commit and push

Comment (faddat) [reply x/ccv/provider/client/proposal_handler.go]: oh thanks a ton <3

**Comment by faddat** [comment]
basically the feedback from @jtremback was quite useful in that he was quite right about the "dirty" branch we started from.  

This branch contains all prior work for only the paths listed above and is otherwise fresh.

File: app/consumer-democracy/app.go [hunk 2]
```diff
@@ -137,7 +135,7 @@ var (
 		ccvdistr.AppModuleBasic{},
 		gov.NewAppModuleBasic(
 			// TODO: eventually remove upgrade proposal handler and cancel proposal handler
-			paramsclient.ProposalHandler, distrclient.ProposalHandler, upgradeclient.ProposalHandler, upgradeclient.CancelProposalHandler,
```
Comment (jtremback) [inline app/consumer-democracy/app.go:140]: Why this removal?

Comment (faddat) [reply app/consumer-democracy/app.go:140]: as of 47 we no longer have a distribution client

**Comment by mpoke** [comment]
> This branch contains the work of five people over nearly a month. Instead of closing the PR when asked for review, as @mpoke did yesterday, please provide guidance if anything isn't to your liking.

@faddat Thank you for your contribution. In order for us to accept contributions, we need contributors to follow our contributing guidelines (see CONTRIBUTING.md in the root of the repo). The reason for this is, while others may submit these pull requests, our team is ultimately responsible for maintaining the code that gets merged and eventually released, meaning that our team needs as much context as possible on the nature of all of the changes being merged.

As such, for large/complex changes, we absolutely require up-front discussion in issues/GitHub discussions. In some cases, like this change, it would be preferable to even have synchronous design sessions to socialize a collective understanding of the nature of incoming changes prior to the work even starting.

Note that if someone from our team is assigned to a particular issue, we will most likely opt to not accept unsolicited PRs that attempt to address those same issues. As a general rule, if someone else is assigned to an issue and you would like to take it over, please reach out to that person and ask whether you can either collaborate with them on it or take it over.

**Comment by mpoke** [comment]
> We feel that it is undesirable to upgrade the repository in the manner described in #852 -- specifically there's little sense in upgrading just one piece of the repository. Let's instead make this module path
> 
> ```go
> module github.com/cosmos/interchain-security/v2
> ```
> 
> And then test v2 against v1 because that is the standard manner of doing these kinds of things in Cosmos.

@faddat Thank you for sharing your concerns. We're happy to discuss them further in the [corresponding opened issue](https://github.com/cosmos/interchain-security/issues/801). In case you missed it, please see [our reply](https://github.com/cosmos/interchain-security/issues/801#issuecomment-1516442358) motivating the reason behind our decision.

**Comment by faddat** [comment]
@mpoke the CI system in use here, does not work for external contributors, or has been disasbled.  I don't know which.  I have created an issue. 

* #873 


* https://github.com/notional-labs/interchain-security/pull/6
* https://github.com/notional-labs/interchain-security/pull/5
* https://github.com/notional-labs/interchain-security/pull/4

**Comment by mpoke** [comment]
> @mpoke the CI system in use here, does not work for external contributors, or has been disasbled. I don't know which. I have created an issue.
> 
> * [CI does not work for external contributors #873](https://github.com/cosmos/interchain-security/issues/873)
> * [update getParams for consumerkeeper and add SlashWithInfractionReason notional-labs/interchain-security#6](https://github.com/notional-labs/interchain-security/pull/6)
> * [More lint notional-labs/interchain-security#5](https://github.com/notional-labs/interchain-security/pull/5)
> * [API change in SDK notional-labs/interchain-security#4](https://github.com/notional-labs/interchain-security/pull/4)

@faddat Please see https://github.com/cosmos/interchain-security/issues/873#issuecomment-1520077285.

**Comment by faddat** [comment]
seems it works again

**Comment by pysel** [comment]
Asked this question in telegram Cosmos Tech channel, was advised to duplicate it here. @mpoke @jtremback, could you please share your thoughts on this?

----
question about params whitelisting in ICS: 

as to my knowledge, whitelisted (in ICS: https://github.com/cosmos/interchain-security/blob/main/app/consumer-democracy/proposals_whitelisting.go#L37) params are not automatically set on chain construction (ref: `NewSimApp`). Hence, I want to ask what we will do with this test https://github.com/cosmos/interchain-security/blob/main/app/consumer-democracy/proposals_whitelisting_test.go#L12 (it initializes a chain, then checks those params are set):

1. Remove the test
2. Set all the whitelisted params prior to test (does not really make sense to me, of course they will be there, it feels like the point of the test disappears)

if at some point I was wrong in this message, correct me please

**Comment by faddat** [comment]
hi @julienbrt, sorry that this is ending up in a repository that you tend; it is possible that informal may learn to use cherry picks, but seems unlikely.  I am cherry picking @vuong177 work to main now

**Comment by faddat** [comment]
@pysel - can you please separate the PR you made into two?

The markdownlint stuff should be targeted to main and then merged into this because it will make reviews easier.

**Comment by pysel** [comment]
hey @faddat, separated, here is the PR with removed test https://github.com/notional-labs/interchain-security/pull/10

**Comment by faddat** [comment]
Thank you so much!

**Comment by faddat** [comment]
Unless anybody has objections, I am going to merge Maine and mark the whole thing as ready for review. Then I would like to schedule a review session with the informal systems team.

**Comment by sontrinh16** [comment]
i think @vuong177 is working to pass the last testcase

**Comment by faddat** [comment]
@pysel I think that we should invite the stewarding team to review this work because this is overall a pretty large pr.

**Comment by faddat** [comment]
@sontrinh16 @vuong177 -- I've got it passing the linter and def want to get the stewarding team's opinion on these changes, so I'll mark ready for review.   If we are failing any unit tests, then we should at this point both:

* consult with the stewarding team
* continue to work on them independently

This is really great work, thanks to everyone who's been grinding on it.

**Comment by faddat** [comment]
@mpoke @jtremback - Can you let us know how you'd like these to be handled?

https://github.com/cosmos/interchain-security/actions/runs/4838431750/jobs/8622881774?pr=869

Usually in these cases, it's a choice between returning an error and handling elsewhere, or setting things to explicitly panic.

**Comment by faddat** [comment]
@vuong177 's recent commit shows signs of #873

**Comment by faddat** [comment]
@jtremback @mpoke I've made it so that legacy parameter changes aren't supported in tests -- @vuong177 @sontrinh16 @pysel can you let me know if you think this is the right plan?  I think it is because of message based gov, but wanted to check

**Comment by faddat** [comment]
> @faddat Thank you for sharing your concerns. We're happy to discuss them further in the https://github.com/cosmos/interchain-security/issues/801. In case you missed it, please see https://github.com/cosmos/interchain-security/issues/801#issuecomment-1516442358 motivating the reason behind our decision.

Reasons keep changing.  Anyhow, Seems that it could make sense to take both paths.  I think that we can get to ideal upgradeability by both modularizing (additional go.mod and go.sum) and using a v2 module path.  The reasoning is that we will want to do upgrades on the 45 and 47 versions over time.

**Comment by faddat** [comment]
@vuong177 

In the sdk that we are using for this work, I noticed that we added the infraction reason to the default slash call.  Can you explain this?

For ease of review I am going to cherry pick your changes for sdk and ibc to main.  Thanks so much

**Comment by sontrinh16** [comment]
> @vuong177
> 
> In the sdk that we are using for this work, I noticed that we added the infraction reason to the default slash call. Can you explain this?
> 
> For ease of review I am going to cherry pick your changes for sdk and ibc to main. Thanks so much
> 
> IBC
> 
> * [changes to support ics ibc-go#3538](https://github.com/cosmos/ibc-go/pull/3538)
> 
> SDK
> 
> * [ics changes against 47 cosmos-sdk#15994](https://github.com/cosmos/cosmos-sdk/pull/15994)

he is working on this, try to test that function out right now

**Comment by sontrinh16** [comment]
> Hi @sontrinh16 in the gva chat, Jehan said that there are not changes needed to IBC,
> 
> "3496 is not needed, we mock out the staking keeper"
> 
> I know that you've requested a call with informal systems, and yes, I believe that it's a good idea to have that call.

huh hmmm, okay let me check this out

**Comment by faddat** [comment]
Ok @sontrinh16 - still hoping we can get a formal review of this work from informal systems.

**Comment by mpoke** [comment]
@faddat please convert this PR to draft as it's still in progress.

**Comment by faddat** [comment]
@mpoke we require review, thank you

**Comment by faddat** [comment]
Thanks so much sir 🙂 Yes, I definitely agree it is time for this to get a review.

**Comment by faddat** [comment]
ummmmmm guys you have moved the e2e test image into a private docker repo

@mpoke @mpoke @jtremback nice collab yo

```
startDocker: 
startDocker: #6 [internal] load metadata for docker.io/informalsystems/hermes:1.2.0
startDocker: #6 CANCELED
startDocker: 
startDocker: #7 [internal] load metadata for docker.io/library/golang:1.19-alpine
startDocker: #7 ERROR: failed to authorize: failed to fetch oauth token: Post "https://auth.docker.io/token": EOF
```

"open" source

**Comment by faddat** [comment]
updated to include

* https://github.com/cosmos/cosmos-sdk/pull/16043

**Comment by faddat** [comment]
@MSalopek please note, earlier comments not addressed at you specifically but instead whole informal team.  Since things aren't checked or replied to in slack, it is important that I figure out how to be heard, as a representative of Notional.  I want to thank you for your work to check our work, it's quite clear that it has caused some progress.

**Comment by faddat** [comment]
@sontrinh16 @vuong177 @pysel - any idea how we can/should handle the gosec items, so we can begin to grind on the integration tests?

**Comment by faddat** [comment]
yep @MSalopek -- very meaningful help, thank you

**Comment by faddat** [comment]
hmmmm, so basically I made the root.go files follow sdk 47 idiom.  However, this resulted in the same problems as before.

**Comment by faddat** [comment]
@sontrinh16 that seems to fix it!  Thank you!  I had just finished up reproducing the issue on my laptop.

**Comment by faddat** [comment]
@sontrinh16 seems that next we need to make some proposal.json files

**Comment by sontrinh16** [comment]
> @sontrinh16 seems that next we need to make some proposal.json files

yepp, i think @vuong177 know how to make it

**Comment by faddat** [comment]
I think I mighta got it.

**Comment by faddat** [comment]
so ok, looks to me like anywhere where we refer to a summary field, we should be referring to a description field

we don't have summaries

**Comment by faddat** [comment]
Well, I guess maybe we need to change description to summary in the json?

I'm a tad confused really

**Comment by faddat** [comment]
* https://github.com/search?q=repo%3Acosmos%2Fcosmos-sdk%20summary&type=code
I wonder if we are dealing with some kind of a bug that will prevent us from ever using a legacy prop?  I don't think legacy props have summaries...

**Comment by mpoke** [comment]
> @mpoke we require review, thank you
> 
> We have also asked for feedback on `gosec` and on `cosmos-sdk` and on `ibc-go` in slack, and on telegram, and here in this PR, and we've gotten none. This PR is ready for review sir.

@faddat thank you for your contribution. With so many changes, it is very difficult to properly review the PR. QA frameworks were broken which meant they could not be relied upon to provide additional context, detect regressions and point to unexpected behavior.

When testing distributed systems, we cannot solely rely on unit tests or integration tests since they don't provide the required level of coverage. MBT and e2e testing approaches provide meaningful test traces to uncover bugs in edge case handling.

Here is a list of issues we found with the current work:
- forks of IBC v7 and SDK v0.47 are used
- differential tests are not functional
- e2e tests are not functional
- the new behavior in `x/staking` is not addressed (can result in provider panics); see SDK PR https://github.com/cosmos/cosmos-sdk/pull/16043
- the `StakingKeeper` interface is incorrectly implemented (SDK v0.47 supports ICS out of the box)

As the SDK upgrade is time sensitive, we will address all these issues in the [feat/upgrade-ics-sdk47-ibc7](https://github.com/cosmos/interchain-security/tree/feat/upgrade-ics-sdk47-ibc7) branch that contains your commits (see https://github.com/cosmos/interchain-security/pull/918). Note that we already fixed most of the issues.

**Comment by sontrinh16** [comment]
> issues

@mpoke thankyou for your review, sorry for the late reply it's has been a busy weekend. I would like to address the issues that you have addressed. 

- Firstly, We have stop using forked version of ibc-go and revert all change related to the cosmos sdk prior to your comment. The only reason why we fork the sdk now is to use the change that was mention here [https://github.com/cosmos/cosmos-sdk/pull/16043](url) and has given credit to @MSalopek where credit is due. Plus we have use some of his commits to stop using forked ibc-go and cosmos-sdk ( you can check in the commits history of this PR )
- Secondly, we have address the new behavior in the `x/staking` before [https://github.com/notional-labs/cosmos-sdk/commit/ed70940b7de6ed7e21f34a3c1bb74bfdf3c8867d](url) but decide to choose @MSalopek approach as mention above
- The `StakingKeeper` interface has been use correct when we stop using forked cosmos-sdk [https://github.com/cosmos/interchain-security/pull/869/commits/e8199e166b926def249af737d2908d987bd958d4](url) again credit to @MSalopek 
- We are working on fixing the e2e tests, as we seen this problem is not our own and many have this too, i have create a PR to try to fix the broken `collect-gentxs` for the main ics-47 PR too [https://github.com/cosmos/interchain-security/pull/926](url)
- As for differential tests are not functional, maybe i miss it somewhere but are there any specify descriptions on this matter, we would love to go through that and learn and try to contribute to the new branch [https://github.com/cosmos/interchain-security/tree/feat/upgrade-ics-sdk47-ibc7](url)

Thankyou for your time, we will continue to support the new PR that @MSalopek has created by create a seperate PRs to that branch as it's our goal too to upgrade ics to sdk-47 so as yours. Good day sir.
